src/status_im/extensions/ethereum.cljs

```diff
@@ -234,12 +234,31 @@
                         :on-failure on-failure}]
      (wrap-with-resolution db json-rpc-args :to execute-send-transaction))))
 
-(defn- parse-log [{:keys [address transactionHash blockHash transactionIndex topics blockNumber logIndex removed data]}]
-  (merge {:data    data
+(defn- event-topic-enc [event params]
+  (let [eventid (str event "(" (string/join "," params) ")")]
+    (abi-spec/sha3 eventid)))
+
+(defn diff-vectors [X Y]
+  (reduce #(remove (fn [x] (= x %2)) %1) X Y))
+
+(defn- decode-data [data hints firstTopic indexed]
+  (let [firstTopicIndex (.indexOf (mapv #(event-topic-enc (% :event) (% :params)) hints) firstTopic)
+        paramsHint (diff-vectors ((get hints firstTopicIndex) :params) indexed)
+        namesHint (mapv keyword ((get hints firstTopicIndex) :names))
+        dataValues (mapv (partial apply str) (partition 64 (string/replace-first data #"0x" "")))]
+    (zipmap namesHint (mapv abi-spec/hex-to-value dataValues paramsHint))))
+
+(defn- decode-topics [topics indexed]
+  (zipmap [:topic1 :topic2 :topic3] (mapv abi-spec/hex-to-value topics indexed)))
+
+(defn parse-log [hints {:keys [address transactionHash blockHash transactionIndex topics blockNumber logIndex removed data]}]
+  (merge {:data data
           :topics  topics
           :address address
           :removed removed}
-        ;; TODO parse data and topics, filter useless solidity first topic, aggregate as events ?
+
+         (when topics {:topics (decode-topics (filterv #(not (= (first topics) %)) topics) ((get hints (.indexOf (mapv #(event-topic-enc (% :event) (% :params)) hints) (first topics))) :indexed))})
```

Comment:
Maybe extract as a function? A bit hard to read. Also is the function call expected? (twice `(`)

Reply:
I was thinking in call that function only one time and save the value in a variable inside of the function, but I didn't find the correct way to do that in this function, could you suggest an example or link to any documentation? Thanks!

Reply:
I would suggest with a simple call each time for now

Reply:
I added a new function and I'm calling it each time, is It ok?  Thanks for your time and feedback :+1: